File: packages/cli/src/server/basic-auth.ts
import type http from "node:http";
import type { Duplex } from "node:stream";
import { secureCompare } from "#app/lib/secure-compare.ts";

export type BasicAuthRequirement = {
  readonly password: string;
  readonly username?: string;
};

export type BasicAuthResult =
  | { readonly ok: true }
  | {
      readonly ok: false;
      readonly reason: "missing" | "malformed" | "invalid";
    };

const CHALLENGE_HEADERS = {
  "content-type": "text/plain; charset=utf-8",
  "www-authenticate": 'Basic realm="proxer"',
};

export const verifyBasicAuthHeader = (
  authorization: string | string[] | undefined,
  requirement: BasicAuthRequirement | undefined,
): BasicAuthResult => {
  if (!requirement) {
    return { ok: true };
  }

  const value = Array.isArray(authorization) ? authorization[0] : authorization;
  if (!value) {
    return { ok: false, reason: "missing" };
  }

  if (!value.toLowerCase().startsWith("basic ")) {
    return { ok: false, reason: "malformed" };
  }

  const decoded = Buffer.from(value.slice("basic ".length), "base64").toString(
    "utf8",
  );
  const separatorIndex = decoded.indexOf(":");
  if (separatorIndex === -1) {
    return { ok: false, reason: "malformed" };
  }

  const username = decoded.slice(0, separatorIndex);
  const password = decoded.slice(separatorIndex + 1);
  if (!secureCompare(requirement.password, password)) {
    return { ok: false, reason: "invalid" };
  }

  if (
    requirement.username !== undefined &&
    !secureCompare(requirement.username, username)
  ) {
    return { ok: false, reason: "invalid" };
  }

  return { ok: true };
};

export const writeBasicAuthChallenge = (
  response: http.ServerResponse,
): void => {
  response.writeHead(401, CHALLENGE_HEADERS);
  response.end("Unauthorized\n");
};

export const writeBasicAuthUpgradeChallenge = (socket: Duplex): void => {
  socket.write(
    "HTTP/1.1 401 Unauthorized\r\n" +
      'www-authenticate: Basic realm="proxer"\r\n' +
      "content-type: text/plain; charset=utf-8\r\n" +
      "connection: close\r\n" +
      "\r\n" +
      "Unauthorized\n",
    () => socket.destroy(),
  );
};

File: packages/cli/lib/src/services/server/basic_auth.dart
import 'dart:convert';
import 'dart:io';

import 'package:proxer/src/services/protocol/frame.dart';

bool _secureCompare(String left, String right) {
  final leftBytes = utf8.encode(left);
  final rightBytes = utf8.encode(right);
  var different = leftBytes.length ^ rightBytes.length;
  final length = leftBytes.length > rightBytes.length
      ? leftBytes.length
      : rightBytes.length;
  for (var index = 0; index < length; index += 1) {
    final a = index < leftBytes.length ? leftBytes[index] : 0;
    final b = index < rightBytes.length ? rightBytes[index] : 0;
    different |= a ^ b;
  }
  return different == 0;
}

bool verifyBasicAuthHeader(
  String? authorization,
  BasicAuthConfig? requirement,
) {
  if (requirement == null) {
    return true;
  }
  if (authorization == null ||
      !authorization.toLowerCase().startsWith('basic ')) {
    return false;
  }
  String decoded;
  try {
    decoded = utf8.decode(base64.decode(authorization.substring(6)));
  } on FormatException {
    return false;
  }
  final separator = decoded.indexOf(':');
  if (separator < 0) {
    return false;
  }
  final username = decoded.substring(0, separator);
  final password = decoded.substring(separator + 1);
  return _secureCompare(requirement.password, password) &&
      (requirement.username == null ||
          _secureCompare(requirement.username!, username));
}

Future<void> writeBasicAuthChallenge(HttpResponse response) async {
  response.statusCode = HttpStatus.unauthorized;
  response.headers
    ..contentType = ContentType.text
    ..set(HttpHeaders.wwwAuthenticateHeader, 'Basic realm="proxer"');
  response.write('Unauthorized\n');
  await response.close();
}
